File: content/13-1d-dynamic-programming/max_product_subarray.js
/**
 * Problem: Maximum Product Subarray
 * GeeksforGeeks: https://practice.geeksforgeeks.org/problems/maximum-product-subarray3604/1
 */

function maxProduct(nums) {
    /**
     * Time Complexity: O(N)
     * Space Complexity: O(1)
     */
    if (!nums || nums.length === 0) return 0;
    
    let res = Math.max(...nums);
    let currMin = 1, currMax = 1;
    
    for (let i = 0; i < nums.length; i++) {
        const num = nums[i];
        if (num === 0) {
            currMin = 1;
            currMax = 1;
            continue;
        }
        
        const temp = currMax * num;
        currMax = Math.max(num * currMax, num * currMin, num);
        currMin = Math.min(temp, num * currMin, num);
        res = Math.max(res, currMax);
    }
    
    return res;
}

if (require.main === module) {
    const assert = require('assert');
    assert.strictEqual(maxProduct([6, -3, -10, 0, 2]), 180);
    assert.strictEqual(maxProduct([-1, -3, -10, 60]), 1800);
    assert.strictEqual(maxProduct([-2, 0, -1]), 0);
    console.log("JS: maxProduct passed tests!");
}

File: content/13-1d-dynamic-programming/max_product_subarray.py
"""
Problem: Maximum Product Subarray
GeeksforGeeks: https://practice.geeksforgeeks.org/problems/maximum-product-subarray3604/1
"""

def max_product(nums: list[int]) -> int:
    """
    Time Complexity: O(N)
    Space Complexity: O(1)
    """
    if not nums:
        return 0
        
    res = max(nums)
    curr_min, curr_max = 1, 1
    
    for num in nums:
        if num == 0:
            curr_min, curr_max = 1, 1
            continue
            
        temp = curr_max * num
        curr_max = max(num * curr_max, num * curr_min, num)
        curr_min = min(temp, num * curr_min, num)
        res = max(res, curr_max)
        
    return res

if __name__ == "__main__":
    assert max_product([6, -3, -10, 0, 2]) == 180
    assert max_product([-1, -3, -10, 60]) == 1800
    assert max_product([-2, 0, -1]) == 0
    print("Python: max_product passed tests!")
